File: events/year_2024/day_01/day_01.py
"""
Advent of Code - Year 2024 - Day 1
https://adventofcode.com/2024/day/1
"""


def preprocessing(puzzle_input: str) -> tuple[list[int], list[int]]:
    """
    Splits the input string into integers, sorts even-indexed and odd-indexed numbers separately,
    and returns them as two lists.
    """
    numbers = [int(item) for item in puzzle_input.split()]
    return sorted(numbers[::2]), sorted(numbers[1::2])


def solver(left_list: list[int], right_list: list[int]) -> tuple[int, int]:
    """
    Calculates the sum of absolute differences between paired elements and the sum of occurrences
    of each element in the right_list list multiplied by its value from the left_list list.
    """
    return (sum(abs(left - right) for left, right in zip(left_list, right_list)),
            sum(left * right_list.count(left) for left in left_list))

File: events/year_2024/day_01/day_01.rs
aoc::main!();

fn preprocessing(input: &str) -> (Vec<i32>, Vec<i32>) {
    input
        .lines()
        .map(|l| l.split_once("   ").unwrap())
        .map(|(l, r)| (l.parse::<i32>().unwrap(), r.parse::<i32>().unwrap()))
        .unzip()
}

fn part_1(lists: (Vec<i32>, Vec<i32>)) -> i32 {
    zip(lists.0.iter().sorted(), lists.1.iter().sorted())
        .map(|(l, r)| (l - r).abs())
        .sum()
}

fn part_2(lists: (Vec<i32>, Vec<i32>)) -> usize {
    lists
        .0
        .iter()
        .counts()
        .iter()
        .map(|(&&l, n)| n * (l as usize) * lists.1.iter().filter(|&&r| l == r).count())
        .sum()
}
